### ISSUE TITLE
Training types in N2D2 (Regression/Reinforcement)

### ISSUE THREAD

**Issue by GasparQ** [CEA-LIST/N2D2#59]
Hi everyone,

I am currently trying to implement the compression neural network from [the **tensorflow_compression** repository](https://github.com/tensorflow/compression) that need to be trained by reinforcement.

This training implies to run 2 optimizers that will minimize a custom loss function and entropy losses.

I was wondering if N2D2 can handle the training of a network with a custom loss function and how hard would it be to implement such a mechanism ?

Last question, can N2D2 run several optimizers at the same time on the same network training ?

**Comment by olivierbichler-cea** [comment]
Hi,
Right now, the loss function in N2D2 is always implicitly defined.
For softmax or logistic, the loss is the cross entropy loss, when used with the "WithLoss=1" parameter. Otherwise, the default loss is the MSE (L2) loss.

The reason why is that the error is defined at the output of a layer with the Cell_Frame[_CUDA]<T>::setOutputTarget() or Cell_Frame[_CUDA]<T>::setOutputTargets(), which set the value of the *input* gradient for the cell to (target - output). These functions are called in the Target class.
So, if you just set the error this way for any cell, the corresponding loss function would be the MSE loss, as the simple difference above is the derivative.
For softmax or logistic however, there is a special parameter "withLoss" that can be enabled, which will simply by-pass the function derivative and directly set the *output* gradient of the function to the difference above. This effectively results to a cross entropy loss with regards to the *input* gradient of these functions, as per the mathematical simplification of the cross entropy loss derivative multiplied by the functions gradient.

So, of course it is possible to define other losses, but its derivative has to be computed explicitly to set the input gradient value of the cell, as there is no automatic differentiation in N2D2. This can be done for example with another Cell_Frame[_CUDA]<T>::setOutputTarget() like function, which would apply the custom loss derivative instead of the (target - output) difference, corresponding to the MSE loss derivative).
Ideally, one would want to refactor a bit this part of N2D2 by directly defining the loss functions and be able to connect it at the output of a cell so that the derivative is computed automatically.

Regarding the use of several optimizers at the same time, I am not sure to understand the setup. But there is no problem in N2D2 to have several targets, and therefore apply different losses to different output cells in the network.

Hope this helps!
I would be glad if we could come up with some refactoring in N2D2 regarding to this issue.

**Comment by Djip007** [comment]
Yes it help!
For the multi optimisation we (GasparQ and Me) need more time to find how it work on tensorflow
Look it is possible to define 2 (or more) Loss... and minimise both... 
But don't know how... look like minimise first... then a second... and the first... ??? 

We have some Brainstorming on possible implementation for Loss function...  need some time to figure if it is good idea... and my english id bad... 
So quickly... in french ;)

Une idée serait de creer les Loss pas combinaison d'element comme pour les Cell: on aurait des propagate/back_propagate mais pas de parametres a optimiser (donc pas de update ... etc). 
par exemple la MSE pourait etre codé avec une ElemWise avec 2 entrées, 1/-1 en "Weights", "Sum" en operation et un **2 en fonction d'activation. l'une des entrées serait la sortie du reseau, l'autre la target (?)  suivit d'une pool "mean" avec une dimension de sortie de [1][1] ...
La composition des 2 propagates permet de calculer l'erreur a minimiser... celle des back_propagate de la propager aux couches précédentes...  
Il suffirait (+ ou - ??) de rajouter un tag "is_loss" pour determiner si cette Layer/Cell doit etre exporter ou ne sert que a l'entrainement... 

Bon reste a voir si ca marche... et si oui... si c'est une bonne idée ;) (et si j'ai reussi a l'expliquer...)
